rfcs/2021-05-23-groups2.md

```diff
@@ -0,0 +1,134 @@
+2) The recipient of this verification message can have fewer messages received. This situation can be resolved in several ways:
+- try to retrieve all messages from the queue that is behind. It might work, but it may be that the sender is simply trying to send messages because the network is down.
+- wait until the next integrity check and only report integrity violation if during the next integrity check they still cannot reconcile the previous integrity check. This is probably an acceptable compromise.
+
+2. Each agent sends message receipts to sending agents - message receipts would contain a signature of the message hashes. In the agent uses the same verification key for each member of the group, as considered below, these receipts can be re-broadcasted to other agents (again, grouping them in one message to avoid `(n-1)*n` messages) as a proof that the messages were delivered.
+
+Comparing with the first approach, there are pros and cons:
+- pros:
+  - the sender would only send such integrity check messages when they have send a message to all parties, thus avoiding the situation when some messages might have been not yet sent (or failed to send and they are retrying).
+- cons:
+  - integrity check message mush contain a signature per member, so it would be substantially larger.
+  - agents must use the same verification key for all members in the group, complicating the group and connections management.
+  - in the same way as in the first scenario, some delivery notifications can be arbitrarily delayed.
+
+With either approach, each agent should probably track all the receipts of all messages from all agents. 
+
+Overall the first approach seems better. It shows who received which messages and the only case where the lack of integrity would be reported if the same message is different or some messages are skipped to some recipients (i.e. they broadcast integrity violation for some of the senders).
+
+## Agent commands and messages syntax
+
+- command `G:gId? NEW` - create group (response is `G:gId OK`, or `ERR` if this group already exists)
+- command `C:cId INTRO G:gId gInfo` - invite existing connection to a group
+- message `C:cId REQ G:invID gInfo` - invitation to join the group
+- command `G:gId? ACPT G:invId` - accept invitation (response is `G:gId OK`)
+- message `G:gId CON C:cId` - 2 connections created with some group member (both for group and direct messages)
+- message `G:gId MEM [C:cId]` - connection created with all group members for a given member or current client
+- command `G:gId SEND msg` - send message to group
+- message `G:gId SENT [C:cId] msgId` - notification that the message is sent and its internal ID
+- message `G:gId MSG C:cId msgdata` - received group message from cId, msgdata is the same set of parameters as in `MSG`
+- command `G:gId ACK msgId` - acknowledge message reception by the client
+- message `G:gId RCVD [C:cId] msgId status` - message delivery notification
```

Comment:
Will this notification be result of integrity check? At least I don't see envelope that support this kind of message

Reply:
The envelope for notifications of message delivery is part of connection protocol, it is not needed separately for groups, as group communication simply uses dedicated connections.